### PR TITLE
feat(topology): add validation for one topology param per storage class

### PR DESCRIPTION
**Pull request by sinhaashish**
1. vi ~/topology/issues/multiple_param_in_sc.yaml
```
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: mayastor-more-than-one-param
parameters:
  ioTimeout: "30"
  protocol: nvmf
  repl: "2"
  nodeSpreadTopologyKey: |
    zone-eu
  nodeHasTopologyKey: |
    zone-us
provisioner: io.openebs.csi-mayastor
volumeBindingMode: Immediate
```

2. vi  ~/topology/issues/multiple_param_in_pvc.yaml
```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: pvc-more-than-one
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 512Mi
  storageClassName: mayastor-more-than-one-param
```
3.  list the pvc and check its not bound.
```
[nix-shell:~/code/rust/mayastor-control-plane]$ kubectl get pvc
NAME                STATUS    VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS                   AGE
pvc-1               Pending                                      mayastor-1                     12m
pvc-more-than-one   Pending                                      mayastor-more-than-one-param   7m22s
```

4. The controller logs says 
``` 
error syncing claim "f51c56df-0854-4748-9168-cf43996b5691": failed to provision volume with StorageClass "mayastor-more-than-one-param": rpc error: code = InvalidArgument desc = Multiple topology parameters are not allowed
```

### CODE REVIEW

File: control-plane/csi-driver/src/context.rs [hunk 1]
```diff
@@ -393,6 +393,17 @@ impl TryFrom<&HashMap<String, String>> for PublishParams {
         )
         .map_err(|_| tonic::Status::invalid_argument("Invalid node_spread_topology_key"))?;
 
+        // add validation for one-to-one mapping of topology params to storage class
+        validate_single_topology_param(
+            pool_affinity_topology_label.clone(),
```
Comment (tiagolobocastro) [inline control-plane/csi-driver/src/context.rs]: Let's pass by reference

File: control-plane/csi-driver/src/context.rs [hunk 2]
```diff
@@ -414,6 +425,31 @@ impl TryFrom<&HashMap<String, String>> for PublishParams {
...
+pub(crate) fn validate_single_topology_param(
+    pool_affinity_topology_label: Option<HashMap<String, String>>,
+    pool_affinity_topology_key: Option<Vec<String>>,
+    pool_has_topology_key: Option<HashMap<String, String>>,
+    node_affinity_topology_label: Option<HashMap<String, String>>,
+    node_affinity_topology_key: Option<Vec<String>>,
+    node_has_topology_key: Option<HashMap<String, String>>,
+    node_spread_topology_key: Option<HashMap<String, String>>,
+) -> Result<(), tonic::Status> {
+    if pool_affinity_topology_label.is_some()
```
Comment (tiagolobocastro) [inline control-plane/csi-driver/src/context.rs]: Am I missing something here, this is just checking if any option is enabled not if more than one are enabled?

Comment (sinhaashish) [reply control-plane/csi-driver/src/context.rs]: Thanks for poiniting this out, rectified it.

File: control-plane/csi-driver/src/context.rs [hunk 3]
```diff
@@ -414,6 +425,33 @@ impl TryFrom<&HashMap<String, String>> for PublishParams {
...
+    number_of_topology_params += node_affinity_topology_key.is_some() as usize;
+    number_of_topology_params += node_has_topology_key.is_some() as usize;
+    number_of_topology_params += node_spread_topology_key.is_some() as usize;
+
+    if number_of_topology_params > 1 {
+        return Err(tonic::Status::invalid_argument(
+            "Multiple topology parameters are not allowed",
+        ));
+    }
+    Ok(())
```
Comment (Abhinandan-Purkait) [inline control-plane/csi-driver/src/context.rs]:
```suggestion
    let topology_params = [
        pool_affinity_topology_label,
        pool_affinity_topology_key,
        pool_has_topology_key,
        node_affinity_topology_label,
        node_affinity_topology_key,
        node_has_topology_key,
        node_spread_topology_key,
    ];

    if topology_params.iter().any(|param| param.is_some()) {
        return Err(tonic::Status::invalid_argument(
            "Topology parameters are not allowed",
        ));
    }
   Ok(())
```
Can we just do this? @tiagolobocastro wdyt?

Comment (Abhinandan-Purkait) [reply control-plane/csi-driver/src/context.rs]: The counting solution is also fine. @sinhaashish upto you.

Comment (Abhinandan-Purkait) [reply control-plane/csi-driver/src/context.rs]: Ohh no, my bad, we need only one, my above solution is incorrect in that case.

```suggestion
   let topology_params = [
        pool_affinity_topology_label,
        pool_affinity_topology_key,
        pool_has_topology_key,
        node_affinity_topology_label,
        node_affinity_topology_key,
        node_has_topology_key,
        node_spread_topology_key,
    ];

    let number_of_topology_params = topology_params.iter().filter(|&&param| param.is_some()).count();

    if number_of_topology_params > 0 {
        return Err(tonic::Status::invalid_argument(
            "Topology parameters are not allowed",
        ));
    }
    Ok(())
```

How about this one?

Comment (sinhaashish) [reply control-plane/csi-driver/src/context.rs]: Out of all these params the affinity keys are vec of strings

so the above code errros out. 
pool_affinity_topology_key,
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `&Option<HashMap<String, String>>`, found `&Option<Vec<String>>`

Comment (sinhaashish) [reply control-plane/csi-driver/src/context.rs]: Updated it , iter() for hashmap and increement for vec of strings

Comment (Abhinandan-Purkait) [reply control-plane/csi-driver/src/context.rs]: You can use is_some() in the list elements to make the list a list of bools, and then check count for true.

Comment (Abhinandan-Purkait) [reply control-plane/csi-driver/src/context.rs]:
```
 let topology_params = [
      pool_affinity_topology_label.is_some(),
      pool_affinity_topology_key.is_some(),
      pool_has_topology_key.is_some(),
      node_affinity_topology_label.is_some(),
      node_affinity_topology_key.is_some(),
      node_has_topology_key.is_some(),
      node_spread_topology_key.is_some(),
  ];

  let number_of_topology_params = topology_params.iter().filter(|&&param| param).count();

  if number_of_topology_params > 1 {
      return Err(tonic::Status::invalid_argument(
          "Multiple topology parameters are not allowed",
      ));
  }

  Ok(())
```

Comment (Abhinandan-Purkait) [reply control-plane/csi-driver/src/context.rs]: Something like this?

Comment (sinhaashish) [reply control-plane/csi-driver/src/context.rs]: done

**Review by Abhinandan-Purkait — APPROVED** [review]

File: control-plane/csi-driver/src/context.rs [hunk 4]
```diff
@@ -414,6 +425,36 @@ impl TryFrom<&HashMap<String, String>> for PublishParams {
     }
 }
 
+/// Validate the topology parameters.
```
Comment (tiagolobocastro) [inline control-plane/csi-driver/src/context.rs]:
```suggestion
/// Ensure there's only 1 topology parameter enabled at a time.
```

**Review by tiagolobocastro — APPROVED** [review]

**Comment by sinhaashish** [comment]
bors merge

**Pull request merged**